### ISSUE TITLE
Update a single node

### ISSUE THREAD

**Issue by abonillasuse** [SUSE/DeepSea#224]
Can we create an option so that a single node is selected for full upgrade? Instead of rolling the entire cluster, we'd like to specify a single node to verify the SP upgrade will do well.

**Comment by BlaineEXE** [comment]
To clarify, are you wanting to target (1) a single, specific node by name regardless of how many roles it has on it, (2) a single instance of a particular role (e.g., mon, rgw), or (3) any single node of DeepSea's choosing?

**Comment by jschmid1** [comment]
@abonillasuse 

The rolling upgrade feature of deepsea takes care of "Will it do well?".

To be precise, it'll stop as soon as it detects _some_ kind of failure.
That is:

* any service that was previously configured is not coming up again. (with configurable timeouts)

* any minion is not reachable anymore

* the cluster health status is not in an acceptable state anymore (with configurable timeouts)

So essentially, it's what you requested, with the addition that it'll just continue to upgrade when the first node upgrades fine.

**Comment by swiftgist** [comment]
@jschmid1 I know that but I understand the initial distrust when any of us say "it just works". :)  

I think the simple solution for this is to create a subdirectory called single or one (or something), copy the default.sls file from /srv/salt/ceph/maintenance/upgrade and strip it of everything except the initial monitor list.  Then change that for loop to execute once.

My advice for installations wanting to use this would be to temporarily add another monitor (likely go from 3 to 4) and then run

salt-run state.orch ceph.maintenance.upgrade.one

That would upgrade one monitor, allow the admin to investigate/understand the process and hopefully trust it.  At the end of the upgrade, remove the extra monitor.  If something does not work for the installation of the one monitor (or more likely the call is to wait until the weekend or scheduled maintenance), then the upgraded monitor can be removed from the policy.cfg and the cluster will be as it was.

**Comment by BlaineEXE** [comment]
I'm not sure that adding a 4th monitor is necessarily advisable unless the user has some spare hardware to throw it on. It also doesn't keep an odd number for the mons' paxos algorithm in the event of some (unlikely) catastrophic failure during the upgrade-test.

Even if we provide the feature of upgrading "one node"--whatever we decide that means to us--what is the user going to do if the upgrade doesn't work? If I were paranoid about the upgrade procedure, which I probably would be, I would make sure I could get back to a healthy cluster as fast as possible in the event of a failure. In a "one node" scenario I'd probably do something like:
> Pick the node that will be the "one node" and make a clonzilla backup of it before updating.

If I'm already going to do the above process, I don't see much different in extending it for the whole upgrade, and indeed, merely upgrading a single monitor to see if it works isn't really a thorough enough test to give me high confidence that the cluster is going to be healthy after the upgrade; there are gateways and osds that can give me grief as well. Even if I were updating "one node" at a time, I'd still extend my above suggestion (on my own production cluster) to be:
> Before the upgrade process, use clonezilla to backup all the OS images. Then begin the upgrade process. DeepSea should fail if anything falls into a nonworking state after the upgrade, and then I can restore a backup to the failed node to keep the cluster healthy, or I can restore all nodes if it's really borked.

I'm not sure if this method is something to be expected from a sysadmin. There may be a more elegant solution, but the gist is that I'd want to be reasonably certain I could revert back to a working state easily in the event of failure. 

After the discussion so far, I'm concerned that this feature doesn't buy the user enough peace of mind to be worth it. On the flip side, it may provide them more peace of mind than is actually merited, and they may stop taking other reasonable precautions.

**Comment by swiftgist** [comment]
Even in proof of concept clusters, the recommendation enforced by the validation requires four nodes.  As far as paxos, an odd number isn't required.  It's the majority.  (I learned that from Joao.)

The reason to add a fourth monitor is precisely to remove the upgraded monitor if something should go wrong.  I doubt there is any method to do a thorough test of upgrading every Ceph component in an automated fashion with hardware constraints.  Upgrading a single node is more about the admin seeing what it looks like and getting an idea of how long it takes.

With Ceph as a distributed system, it may technically be near impossible to guarantee that every byte on every node of the cluster can be reverted to the same previous state without shutting down all the nodes.  

As far as precautions, I believe all have a development/staging cluster that are running a production cluster.  Those that do will have backups of the production servers, but not necessarily the development/staging cluster.  Going forward to the previous configuration is essentially the closest we can simulate reverting to a previous state.

**Comment by BlaineEXE** [comment]
Okay, so moving forward, I've been trying to figure out how to pass arguments to runners. I think I can do so by adding custom pillar data in with the request, but I've been having trouble testing it successfully. Is this reasonably something I should continue trying to do (passing arguments), or should I refocus in another direction?

**Comment by jschmid1** [comment]
I still think that we don't need this feature. The upgrading process will _NEVER_ bring the cluster down or leave it in a critical state, given you followed our recommendations regarding mon & osd count. We should rather focus on explaining the recommendations/requirements and emphasize on the strategy we took to guarantee the uptime.

If they really want to understand the process, one can enable the 'debug' flag in

/srv/salt/ceph/upgrade/default.sls

which will add a `--dry-run` to zypper and disable reboots (atleast those triggered by packagemanager.py)

Even if you have the feature of upgrading only one node. What node to take? rgw? iscsi? mon? Will they behave the same? 
Imho that just lulls you into a sense of security..

**Comment by BlaineEXE** [comment]
@jschmid1, I think we are on the same page about this being a false sense of security without substantial benefit.

**Comment by abonillasuse** [comment]
I'm after #1 @blaineexe the goal is to ensure the hardware can survive a service pack upgrade, that the controller won't be a problem, etc. I'm not worried about SES but mostly the whole upgrade of the OS and all it implies.

**Comment by jschmid1** [comment]
So, If you go with the Deepsea recommended way of upgrading your cluster and it fails on the first attempt ( your first node actually died because of reasons ), the process will stop. You will end up in the exact same spot as you'd have been after upgrading only _one_ node.

**Comment by BlaineEXE** [comment]
I'm still with @jschmid1. We already have the functionality needed to ensure the hardware can survive an os distro upgrade. Adding this functionality will only add complexity without adding anything new. Unless someone can provide a compelling user story that suggests this is useful beyond our current upgrade process, I would consider this a wontfix.

@abonillasuse, can you provide a user story for us? How do you see this working?

**Comment by abonillasuse** [comment]
@blaineexe every single customer would want to upgrade one single system before all others. In all current storage implementations, we update a head and then another. This is going to be necessary as well when we have SLE15 because there are hardware updatability changes and we cannot afford to test on the entire cluster in parallel.

**Comment by swiftgist** [comment]
@abonillasuse Just so you know, the rolling upgrade will not do updates to the nodes in parallel.  That is what Joshua explained early on.  If something (including Ceph processes outside the upgraded node) has failed, the upgrade stops.

The only difference I can see between what you are requesting and what is currently provided is whether we add a switch to only do the first node.  My impression of the benefit is for the admin to see the length of time it takes to update a node and to your point, exercise the OS update process to make sure repos are configured properly and feel better about the OS update in general.  The next step would be making any corrections and finally, running the entire upgrade.

Beyond that, if you find an incompatibility that we missed, the upgrade will stop with that single node down.  In general, I would expect that to be the first node of a particular type.

**Comment by smithfarm** [comment]
It should be possible to upgrade the OS of any node (or nodes) *without* using DeepSea, first, and run the DS rolling upgrade later to finish the upgrade. Correct?

Assuming the rolling upgrade code in DS will not see a partially upgraded node as an issue, that should address @abonillasuse 's use case.

**Comment by jschmid1** [comment]
Right, @smithfarm. The rolling upgrade is idempotent and will just noop an already updated node.

**Comment by swiftgist** [comment]
Upgrades are addressed and other pieces are out of scope for DeepSea.
